core/config/config_helper.hpp

```diff
@@ -315,6 +316,32 @@ inline std::shared_ptr<typename Csr::strategy_type> get_strategy(
 }
 
 
+/**
+ * This is a decroator on top of config node. When we try to get a node by a
+ * key, it registers the keys as the allowed_keys. When this object is
+ * destroyed, it will check whether the config node only contains the key in
+ * allowed_keys. If some key is not in allowed_keys, it will throw an exception.
+ */
+class config_decorator {
```

Comment:
I'm trying to look for a better name for this class - while it is a decorator, it should probably have a more specific name, maybe `config_check_decorator`?